extensions/amp-base-carousel/1.0/base-carousel.js

```diff
@@ -0,0 +1,62 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import * as Preact from '../../../src/preact';
+import {Arrow} from './arrow';
+import {Scroller} from './scroller';
+import {useRef, useState} from '../../../src/preact';
+
+/**
+ * @param {!JsonObject} props
+ * @return {PreactDef.Renderable}
+ */
+export function BaseCarousel(props) {
+  const {style, arrowPrev, arrowNext, children, currentSlide, loop} = props;
+  const {0: curSlide, 1: setCurSlide} = useState(currentSlide || 0);
+  const ignoreProgrammaticScroll = useRef(true);
+  const setRestingIndex = (i) => {
+    ignoreProgrammaticScroll.current = true;
+    setCurSlide(i);
+  };
+
+  return (
+    <div style={style}>
+      <Scroller
+        ignoreProgrammaticScroll={ignoreProgrammaticScroll}
+        loop={loop}
+        restingIndex={curSlide}
+        setRestingIndex={setRestingIndex}
+      >
+        {children}
+      </Scroller>
+      <Arrow
+        customArrow={arrowPrev}
+        dir={-1}
+        length={children.length}
+        loop={loop}
+        restingIndex={curSlide}
+        setRestingIndex={setRestingIndex}
+      ></Arrow>
+      <Arrow
+        customArrow={arrowNext}
+        dir={1}
+        length={children.length}
```

Comment:
This is quite a bit info to pass to an arrow. It feels like it should instead just be a single `onClick` and we'd do all this logic with looping, children length here instead. WDYT?

Reply:
This is fine. I would still have to pass `customArrow`, `onClick`, `disabled`, and `dir` (to render the appropriate direction). We could eventually get rid of `dir` by creating `ArrowPrev` and `ArrowNext` classes which pass `dir` to `Arrow`, though if that is desirable I'd prefer to make that change in a subsequent PR where we actually replace the `<<` text arrow with the SVGs.